### ISSUE TITLE
Leader Election Extension

### ISSUE THREAD

**Issue by skhalash** [open-telemetry/opentelemetry-collector-contrib#34460]
### The purpose and use-cases of the new component

There can be a use case where one needs to run components in an HA mode. But currently running multiple instances of components like  `k8sclusterreceiver` would
cause collection of duplicate metrics. In order to be able to run such receivers in an HA mode one needs to make sure only one receiver is active. This can be achieved by utilizing the leader election mechanism provided by Kubernetes client library. 

The support for extension can be implemented as an extension so that any component that wants to leverage leader election can use it easily.

### Example configuration for the component

```yaml
receivers:
  k8sclusterreceiver:
    leader_elector: leader_elector
extensions:
  leader_elector:
    auth_type: kubeConfig
    lease_name: foo
    lease_namespace: default
```

### Telemetry data types supported

traces, metrics, and logs

### Is this a vendor-specific component?

- [ ] This is a vendor-specific component
- [ ] If this is a vendor-specific component, I am a member of the OpenTelemetry organization.
- [ ] If this is a vendor-specific component, I am proposing to contribute and support it as a representative of the vendor.

### Code Owner(s)

@skhalash @a-thaler @rakesh-garimella 

### Sponsor (optional)

@dmitryax ([volunteered](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/34460#issuecomment-2590714726))

### Additional context

I work at SAP on a project called Kyma: https://kyma-project.io/#/. In Kyma, we recently developed such a receiver and we are already testing it out in a production setup: https://github.com/kyma-project/opentelemetry-collector-components/tree/main/receiver/singletonreceivercreator.

We’ve noticed discussions in the community about introducing leader election in the k8sobjects and k8scluster receivers. That's why we believe that a generic mechanism could be quite beneficial. If there’s interest, we are ready to contribute it to the Open Telemetry project.

**Comment by skhalash** [comment]
Hey @ChrsMark! Thanks for the feedback. Yes, exactly the leader is responsible for enabling a sub-receiver (such as `k8sclusterreceiver` and `k8sobjectsreceiver`) if the collector is running as a DaemonSet or a Deployment.

Yes, I fully agree with the concerns about resource limits/requests. However, as you said, it should be properly documented. Not sure if there is a way to workaround it.

Regarding putting some extra load on the k8s API - do you mean querying/updating leases?

**Comment by ChrsMark** [comment]
> Regarding putting some extra load on the k8s API - do you mean querying/updating leases?

@skhalash yes, but this is something that can also be properly documented along with some perf tests results so as users be aware of any possible impact to their clusters.

**Comment by dmitryax** [comment]
Why does it have to be a separate receiver? I think this should be an extension providing an interface that _any_ receiver can connect to check if it's a leader (have to do the work) or not (do nothing).

**Comment by skhalash** [comment]
Hey @dmitryax! Thanks so much for your response! We’d be happy to contribute such a component. I don’t have much experience with extensions—would this mean that every receiver needing this functionality would require code modifications? Implementing it as a delegating receiver, similar to receiver-creator, could avoid it.

**Comment by dmitryax** [comment]
Modifying code is fine - better than having two similar receivers. It's better to write some code, but keep the user interface clean. Restricting it to receiver_creator, we won't be able to use leader election for receivers that don't work with receiver creator, for example https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/k8seventsreceiver, or don't need to use it all, e.g. collecting data from a static endpoint. Also, future of receiver_creator is unclear. We might take a different approach to discovery and consolidate all scraping receivers in one, see https://github.com/open-telemetry/opentelemetry-collector/issues/11238

**Comment by skhalash** [comment]
Thanks for the response! We'll explore the idea of implementing leader election as an extension instead of a receiver and will get back to you.

**Comment by skhalash** [comment]
Hey @dmitryax,

We're exploring a specific use case and wanted to get your thoughts. The scenario involves multiple receivers within a single running collector instance that may need to operate in singleton mode, using leader election. To support this, each receiver might need its own lease, allowing each to be managed by a different leader.

We're considering two possible approaches:

1. Embedding the lease configuration directly within each interested receiver’s configuration.
2. Creating multiple instances of a new extension, like `singleton/foo` and `singleton/bar`, each referenced by its respective receiver.

The second approach seems simpler to implement. However, from what I’ve seen, there aren’t any examples in the wild of multiple instances of extensions being registered. Do you think this approach would be feasible?

**Comment by JaredTan95** [comment]
This is a very interesting feature, and many of the plugins in otel col contrib are currently restricted to running on a single instance.

**Comment by skhalash** [comment]
We did some investigation and agree that implementing leader election as an extension is indeed a cleaner solution. We will provide a PR with the implementation soon.

**Comment by chenlujjj** [comment]
hello @skhalash wondering what's the progress there ?

**Comment by rakesh-garimella** [comment]
Hi @chenlujjj  I am actively working on this. I plan to create a PR with initial implementation of the extension in this week.

**Comment by rakesh-garimella** [comment]
Initial Implementation for the extension has been done [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37144). @dmitryax please have a look.

**Comment by dmitryax** [comment]
I sopnsor this component

**Comment by rakesh-garimella** [comment]
First PR is created here [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37266)

**Comment by rakesh-garimella** [comment]
The second PR with actual implementation is here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38015. So please reopen the issue
